<!--
  This file will launch rViz along with the mmWave sensor node and configure a TI mmWave 6843 sensor using a 3D config
 -->

<launch>

    <!-- Input arguments -->
    <arg name="device" value="6843" doc="TI mmWave sensor device type [1443, 1642]"/>
    <arg name="config" value="3d_Tracking" doc="TI mmWave sensor device configuration [3d_best_range_res (not supported by 1642 EVM), 2d_best_range_res]"/>
    <arg name="max_allowed_elevation_angle_deg" default="90" doc="Maximum allowed elevation angle in degrees for detected object data [0 > value >= 90]}"/>
    <arg name="max_allowed_azimuth_angle_deg" default="90" doc="Maximum allowed azimuth angle in degrees for detected object data [0 > value >= 90]}"/>

    <arg name="command_port" default="/dev/ttyXRUSB0"/>
    <arg name="data_port" default="/dev/ttyXRUSB1"/>

    <!-- mmWave_Manager node -->
    <node pkg="ti_mmwave_tracker_rospkg" type="ti_mmwave_tracker_rospkg" name="ti_mmwave" output="screen">
        <param name="command_port" value="$(arg command_port)"  />
        <param name="command_rate" value="115200"   />
        <param name="data_port" value="$(arg data_port)"  />
        <param name="data_rate" value="921600"   />
        <param name="max_allowed_elevation_angle_deg" value="$(arg max_allowed_elevation_angle_deg)"   />
        <param name="max_allowed_azimuth_angle_deg" value="$(arg max_allowed_azimuth_angle_deg)"   />
        <param name="frame_id" value="ti_mmwave"/>
        <param name="mmWaveCLI_name" value="mmWaveCLI" />

        <!-- Way too many remaps - this is a hack to fix namespacing -->
        <remap from="/ti_mmwave/radar_scan_pcl" to="ti_mmwave/radar_scan_pcl"/>
        <remap from="/ti_mmwave/radar_scan" to="ti_mmwave/radar_scan"/>
        <remap from="/ti_mmwave/radar_scan_markers" to="ti_mmwave/radar_scan_markers"/>
        <remap from="/ti_mmwave/radar_trackarray" to="ti_mmwave/radar_trackarray"/>
        <remap from="/ti_mmwave/radar_track" to="ti_mmwave/radar_track"/>
        <remap from="/ti_mmwave/radar_trackid" to="ti_mmwave/radar_trackid"/>
    </node>

    <!-- mmWaveQuickConfig node (terminates after configuring mmWave sensor) -->
    <node pkg="ti_mmwave_tracker_rospkg" type="mmWaveQuickConfigTracker" name="ti_mmwave_config" args="$(find ti_mmwave_tracker_rospkg)/cfg/AOP_3d_Tracking.cfg" output="screen" >
        <param name="mmWaveCLI_name" value="mmWaveCommSrv/mmWaveCLI" />
    </node>

    <!--
    <node pkg="tf" type="static_transform_publisher" name="radar_baselink_1" args="5 0 0 0 0 0 ti_mmwave_pcl ti_mmwave 100"/>
    -->
</launch>
